Dockerfile

```diff
@@ -1,31 +1,62 @@
-FROM php:7.2-fpm
+ARG COMPOSER_IMAGE_VERSION=1.10
+ARG NODE_IMAGE_VERSION=14.4
+ARG PHP_IMAGE_VERSION=7.2-fpm-buster
+ARG NGINX_PACKAGE_VERSION=1.14.2-2+deb10u1
+ARG SUPERVISOR_PACKAGE_VERSION=3.3.5-1
+ARG CRON_PACKAGE_VERSION=3.0pl1-134+deb10u1
 
-# Install system packages (Git, packages for archives, Node.js)
-RUN apt-get update \
-    && apt-get install -y git libzip-dev zlib1g-dev unzip \
-    && curl -sL https://deb.nodesource.com/setup_14.x | bash - && apt-get install -y nodejs \
-    && rm -rf /var/lib/apt/lists/*
+FROM composer:${COMPOSER_IMAGE_VERSION} as composer
 
-# Install PHP extensions
-RUN docker-php-ext-install pdo_mysql && docker-php-ext-install zip && docker-php-ext-install calendar
+FROM node:${NODE_IMAGE_VERSION} as assets
+
+WORKDIR /build
+
+COPY yarn.lock package.json webpack.mix.js ./
+
+RUN yarn install
+
+COPY ./resources ./resources
+
+RUN yarn run production
 
-# Ensure commands are running successfully before continuing
-SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+FROM php:${PHP_IMAGE_VERSION}
 
-# Install Composer
-RUN curl -sS https://getcomposer.org/installer | php -- --install-dir=/usr/local/bin --filename=composer
+ARG NGINX_PACKAGE_VERSION
+ARG SUPERVISOR_PACKAGE_VERSION
+ARG CRON_PACKAGE_VERSION
 
-# Install Yarn
-RUN npm install -g yarn
+# Install nginx as webserver, and some packages needed for php extensions
+RUN apt-get update && apt-get install -y \
+    nginx-light=${NGINX_PACKAGE_VERSION} \
+    supervisor=${SUPERVISOR_PACKAGE_VERSION} \
+    cron=${CRON_PACKAGE_VERSION} \
+    libzip-dev \
+    unzip \
+ && rm -rf /var/lib/apt/lists/*
+
+# Install PHP extensions
+RUN docker-php-ext-install pdo_mysql \
+ && docker-php-ext-install zip \
+ && docker-php-ext-install calendar
+
+COPY nginx.conf /etc/nginx/sites-enabled/default
+COPY cron.conf /etc/cron.d/budget
+COPY supervisord.conf /etc/supervisor/conf.d/budget.conf
+COPY docker_boot.sh /usr/bin/docker-entrypoint
+ENTRYPOINT ["docker-entrypoint"]
+
+COPY --from=composer /usr/bin/composer /usr/bin/composer
 
 WORKDIR /usr/share/nginx/budget
-COPY . .
 
-# Required before we're able to run any "php artisan" commands
-RUN composer install
+COPY composer.json composer.lock ./
+RUN composer install --no-dev --no-scripts --no-autoloader
+
+COPY --chown=www-data . ./
+COPY --chown=www-data --from=assets /build/public/ ./public/
 
-RUN php artisan budget:install
+RUN composer dumpautoload -oa
```

Comment:
Why is this required after `composer install`?

Reply:
Because I didn't dump the autoloader during installation with ```--no-autoloader```.

Initially I only copy the composer.json and .lock files into the container and run the installation.   
This way that layer won't get rebuild if some changes in the actual source code is made; it will only rebuild if the actual composer files get changed.

Reply:
Gotcha, smart.